processor/datadogprocessor/agent.go

```diff
@@ -0,0 +1,186 @@
+	// disable the HTTP receiver
+	cfg.ReceiverPort = 0
+	// set the API key to succeed startup; it is never used nor needed
+	cfg.Endpoints[0].APIKey = "skip_check"
+	// set the default hostname to the tranlator's placeholder; in the case where no hostname
+	// can be deduced from incoming traces, we don't know the default hostname (because it is set
+	// in the exporter). In order to avoid duplicating the hostname setting in the processor and
+	// exporter, we use a placeholder and fill it in later (in the Datadog Exporter or Agent OTLP
+	// Ingest). This gives a better user experience.
+	cfg.Hostname = translator.UnsetHostnamePlaceholder
+	pchan := make(chan *api.Payload, 1000)
+	a := agent.NewAgent(ctx, cfg)
+	// replace the Concentrator (the component which computes and flushes APM Stats from incoming
+	// traces) with our own, which uses the 'out' channel.
+	a.Concentrator = stats.NewConcentrator(cfg, out, time.Now())
+	// ...and the same for the ClientStatsAggregator; we don't use it here, but it is also a source
+	// of stats which should be available to us.
+	a.ClientStatsAggregator = stats.NewClientStatsAggregator(cfg, out)
+	// lastly, start the OTLP receiver, which will be used to introduce ResourceSpans into the traceagent,
+	// so that we can transform them to Datadog spans and receive stats.
+	a.OTLPReceiver = api.NewOTLPReceiver(pchan, cfg)
+	return &traceagent{
+		Agent: a,
+		exit:  make(chan struct{}),
+		pchan: pchan,
+	}
+}
+
+// Start starts the traceagent, making it ready to ingest spans.
+func (p *traceagent) Start() {
+	// we don't need to start the full agent, so we only start a set of minimal
+	// components needed to compute stats:
+	for _, starter := range []interface{ Start() }{
+		p.Concentrator,
+		p.ClientStatsAggregator,
+		// we don't need the samplers' nor the processor's functionalities;
+		// but they are used by the agent nevertheless, so they need to be
+		// active and functioning.
+		p.PrioritySampler,
+		p.ErrorsSampler,
+		p.NoPrioritySampler,
+		p.EventProcessor,
+	} {
+		starter.Start()
+	}
+
+	p.goDrain()
+	p.goProcess()
+}
+
+// Stop stops the traceagent, making it unable to ingest spans. Do not call Ingest after Stop.
+func (p *traceagent) Stop() {
+	for _, stopper := range []interface{ Stop() }{
+		p.Concentrator,
+		p.ClientStatsAggregator,
+		p.PrioritySampler,
+		p.ErrorsSampler,
+		p.NoPrioritySampler,
+		p.EventProcessor,
+	} {
+		stopper.Stop()
+	}
+	close(p.exit)
+	p.wg.Wait()
+}
+
+// goDrain drains the TraceWriter channel, ensuring it won't block. We don't need the traces,
+// nor do we have a running TraceWrite. We just want the outgoing stats.
+func (p *traceagent) goDrain() {
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+	loop:
+		for {
+			select {
+			case <-p.TraceWriter.In:
+				// we don't write these traces anywhere; drain the channel
+			case <-p.exit:
+				break loop
+			}
```

Comment:
Why not just return? Labeled loops are uncommon in this codebase

Reply:
I guess it was habit. It's actually a common practice in Go, and less error prone. If someone decides to ever add logic after the loop, it may be surprising to exit it at such a deep indentation level.

I have changed to returns though as I don't feel strongly on this.

Reply:
I don't feel strongly about either way, but I think we should err on the side of consistency so that other contributors can make changes easily if needed